In [1]:
import pandas as pd
import numpy as np
import pickle
from datetime import datetime

In [2]:
dtrader_1=np.load('alphastock/code/RL/ccso_result/2023_01_01-2023_03_31.npy')
dtrader_1=pd.DataFrame(dtrader_1)
dtrader_1.columns=['daily_return']
dtrader_return_1=dtrader_1/dtrader_1.shift(1)
dtrader_return_1.fillna(1,inplace=True)

dtrader_2=np.load('alphastock/code/RL/ccso_result/2023_04_01-2023_06_30.npy')
dtrader_2=pd.DataFrame(dtrader_2)
dtrader_2.columns=['daily_return']
dtrader_return_2=dtrader_2/dtrader_2.shift(1)
dtrader_return_2.fillna(1,inplace=True)

dtrader_3=np.load('alphastock/code/RL/ccso_result/2023_07_01-2023_09_30.npy')
dtrader_3=pd.DataFrame(dtrader_3)
dtrader_3.columns=['daily_return']
dtrader_return_3=dtrader_3/dtrader_3.shift(1)
dtrader_return_3.fillna(1,inplace=True)

dtrader_4=np.load('alphastock/code/RL/ccso_result/2023_10_01-2023_12_31.npy')
dtrader_4=pd.DataFrame(dtrader_4)
dtrader_4.columns=['daily_return']
dtrader_return_4=dtrader_4/dtrader_4.shift(1)
dtrader_return_4.fillna(1,inplace=True)

In [3]:
dtrader_return = pd.concat([dtrader_return_1, dtrader_return_2, dtrader_return_3, dtrader_return_4])

In [4]:
dtrader_return

daily_return
0       1.000000
1       1.000000
2       1.013637
3       0.997379
4       0.999823
..           ...
55      1.002194
56      0.997849
57      1.008913
58      1.017788
59      1.004746

[242 rows x 1 columns]

In [5]:
df_return=pd.DataFrame()
dtrader_return=dtrader_return/dtrader_return.shift(1)
dtrader_return.fillna(1,inplace=True)
df_return['daily_return']=dtrader_return['daily_return']-1

In [6]:
'''
df_return=pd.DataFrame()
dtrader=np.load('alphastock/code/RL/res/best_test_cr_whole.npy')
dtrader=dtrader.reshape(248)
dtrader=pd.DataFrame(dtrader)
dtrader.columns=['daily_return']
dtrader_return=dtrader/dtrader.shift(1)
dtrader_return.fillna(1,inplace=True)
df_return['daily_return']=dtrader_return['daily_return']-1
'''

"\ndf_return=pd.DataFrame()\ndtrader=np.load('alphastock/code/RL/res/best_test_cr_whole.npy')\ndtrader=dtrader.reshape(248)\ndtrader=pd.DataFrame(dtrader)\ndtrader.columns=['daily_return']\ndtrader_return=dtrader/dtrader.shift(1)\ndtrader_return.fillna(1,inplace=True)\ndf_return['daily_return']=dtrader_return['daily_return']-1\n"

In [7]:
'''
#大盘和日收益率
path='TRA/data/splab.pkl'
f=open(path,'rb')
index_df=pickle.load(f)
index_df = index_df.reset_index()
# index_df.rename(columns={'label':'daily_return', 'dt':'datetime'},inplace=True)
index_df.rename(columns={'label':'daily_return', 'Date':'datetime'},inplace=True)
index_df=index_df[index_df['datetime']>'2022-12-31']
index_df.reset_index(inplace=True,drop=True)
index_df = index_df.groupby('datetime')['daily_return'].mean().reset_index()
index_date_unique=index_df['datetime'].unique()
'''
csitrade=pd.read_csv('E:/baselines/day_level_dataset/ccso.csv')
csitrade['close_r']=csitrade['close']/csitrade['close'].shift(1)
csitrade.reset_index(inplace=True,drop=True)
df_trade=csitrade
df_trade.rename(columns={'kdcode':'instrument','dt':'datetime'},inplace=True)
df_trade['datetime'] = df_trade['datetime'].astype('datetime64[ns]')
df_trade=df_trade[df_trade['datetime']>'2023-01-01']
df_trade.rename(columns={'close_r' : 'daily_return'}, inplace=True)
df_trade['daily_return'] = df_trade['daily_return'] - 1
index_df = df_trade.groupby('datetime')['daily_return'].mean().reset_index()

C:\Users\胡逸凡\AppData\Local\Temp\ipykernel_4832\3311902912.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_trade.rename(columns={'close_r' : 'daily_return'}, inplace=True)
C:\Users\胡逸凡\AppData\Local\Temp\ipykernel_4832\3311902912.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_trade['daily_return'] = df_trade['daily_return'] - 1


In [8]:
portfolio_df_performance = df_return
index_df_performance = index_df.set_index(['datetime'])

In [9]:
# ARR = (portfolio_df_performance['daily_return']+1).cumprod()
# ARR.to_csv("alphastock.csv", index=None)

In [10]:
portfolio_df_performance

daily_return
0   0.000000e+00
1  -1.110223e-16
2   1.363703e-02
3  -1.603888e-02
4   2.450352e-03
..           ...
55  6.097642e-04
56 -4.335360e-03
57  1.108752e-02
58  8.796324e-03
59 -1.281376e-02

[242 rows x 1 columns]

In [11]:
index_df_performance

daily_return
datetime                
2023-01-03      0.012240
2023-01-04     -0.000450
2023-01-05      0.007891
2023-01-06      0.001752
2023-01-09      0.003574
...                  ...
2023-12-25      0.000425
2023-12-26     -0.007765
2023-12-27      0.004482
2023-12-28      0.018146
2023-12-29      0.003977

[242 rows x 1 columns]

In [12]:
alpha_df_performance = pd.DataFrame()
alpha_df_performance['portfolio_daily_return'] = portfolio_df_performance['daily_return']
alpha_df_performance['index_daily_return'] = index_df_performance['daily_return']
alpha_df_performance['alpha_daily_return'] = alpha_df_performance['portfolio_daily_return'] - \
                                                alpha_df_performance['index_daily_return']
alpha_df_performance['portfolio_net_value'] = (alpha_df_performance['portfolio_daily_return'] + 1).cumprod()
alpha_df_performance['index_net_value'] = (
        alpha_df_performance['index_daily_return'] + 1).cumprod()
alpha_df_performance['alpha_net_value'] = (alpha_df_performance['alpha_daily_return'] + 1).cumprod()

net_value_columns = ['portfolio_net_value',
                        'index_net_value',
                        'alpha_net_value']

alpha_statistics_df = pd.DataFrame(index=alpha_df_performance[net_value_columns].columns,
                                    columns=["年化收益", "年化波动率", "最大回撤率", "夏普率", "Calmar", "IR", "月度胜率"])

# alpha_df_performance.set_index("dt", inplace=True)
alpha_df_performance.index = pd.to_datetime(alpha_df_performance.index)
monthly_statistics_df = alpha_df_performance[net_value_columns].resample('m').last()
monthly_statistics_df = pd.concat([alpha_df_performance[:1][
                                        ['portfolio_net_value',
                                        'index_net_value', 'alpha_net_value']],
                                    monthly_statistics_df])
monthly_statistics_df = monthly_statistics_df.pct_change()
monthly_statistics_df = monthly_statistics_df.dropna()
monthly_statistics_df.index = monthly_statistics_df.index.date
## TODO 补充第一年的数据
yearly_statistics_df = alpha_df_performance[net_value_columns].resample('y').last()
yearly_statistics_df = pd.concat([alpha_df_performance[:1][
                                        ['portfolio_net_value',
                                        'index_net_value', 'alpha_net_value']],
                                    yearly_statistics_df])
yearly_statistics_df = yearly_statistics_df.pct_change()
yearly_statistics_df = yearly_statistics_df.dropna()
yearly_statistics_df.index = yearly_statistics_df.index.date

alpha_statistics_df.loc[:, "年化收益"] = np.mean(
    (alpha_df_performance[net_value_columns].tail(1)) ** (252 / len(alpha_df_performance)) - 1)
alpha_statistics_df.loc[:, "年化波动率"] = np.std(
    alpha_df_performance[net_value_columns] / alpha_df_performance[net_value_columns].shift(1) - 1) * np.sqrt(
    252)
alpha_statistics_df.loc[:, "累积收益"] = np.mean(alpha_df_performance[net_value_columns].tail(1) - 1)
alpha_statistics_df.loc[:, "累积波动率"] = np.std(
    alpha_df_performance[net_value_columns] / alpha_df_performance[net_value_columns].shift(1) - 1)
alpha_statistics_df.loc[:, "最大回撤率"] = np.min(
    (alpha_df_performance[net_value_columns] - alpha_df_performance[net_value_columns].cummax()) /
    alpha_df_performance[net_value_columns].cummax())
alpha_statistics_df.loc[:, "夏普率"] = alpha_statistics_df["年化收益"] / alpha_statistics_df["年化波动率"]
alpha_statistics_df.loc[:, "Calmar"] = alpha_statistics_df["年化收益"] / abs(alpha_statistics_df["最大回撤率"])
alpha_statistics_df.loc[:, "IR"] = np.mean(
    alpha_df_performance[net_value_columns] / alpha_df_performance[net_value_columns].shift(1) - 1) * np.sqrt(
    252) / np.std(alpha_df_performance[net_value_columns] / alpha_df_performance[net_value_columns].shift(1) - 1)
alpha_statistics_df.loc[:, "月度胜率"] = monthly_statistics_df[
                                            monthly_statistics_df > 0].count() / monthly_statistics_df.count()

C:\Users\胡逸凡\AppData\Local\Temp\ipykernel_4832\4043393616.py:20: FutureWarning: 'm' is deprecated and will be removed in a future version, please use 'ME' instead.
  monthly_statistics_df = alpha_df_performance[net_value_columns].resample('m').last()
C:\Users\胡逸凡\AppData\Local\Temp\ipykernel_4832\4043393616.py:25: FutureWarning: The default fill_method='pad' in DataFrame.pct_change is deprecated and will be removed in a future version. Either fill in any non-leading NA values prior to calling pct_change or specify 'fill_method=None' to not fill NA values.
  monthly_statistics_df = monthly_statistics_df.pct_change()
C:\Users\胡逸凡\AppData\Local\Temp\ipykernel_4832\4043393616.py:29: FutureWarning: 'y' is deprecated and will be removed in a future version, please use 'YE' instead.
  yearly_statistics_df = alpha_df_performance[net_value_columns].resample('y').last()
C:\Users\胡逸凡\AppData\Local\Temp\ipykernel_4832\4043393616.py:34: FutureWarning: The default fill_method='pad' in DataFrame.pct_

In [13]:
alpha_statistics_df

年化收益    年化波动率     最大回撤率       夏普率    Calmar       IR  \
portfolio_net_value  0.004943  0.16387 -0.040244  0.030161  0.122814  0.11212   
index_net_value      0.004943      NaN -0.040244       NaN  0.122814      NaN   
alpha_net_value      0.004943      NaN -0.040244       NaN  0.122814      NaN   

                    月度胜率      累积收益     累积波动率  
portfolio_net_value  NaN  0.004746  0.010323  
index_net_value      NaN  0.004746       NaN  
alpha_net_value      NaN  0.004746       NaN

alpha stock 

ASR = ARR / AVol;   CR = ARR / abs(MDD)
        ARR    AVol    MDD    ASR    CR    IR
hs300 -0.164  0.153  -0.245 -1.072 -0.669 -1.098

zz500 -0.017  0.148  -0.166 -0.115 -0.102 -0.043

sp500  0.122  0.140  -0.126  0.871  0.968  0.892
   
nas100 0.372  0.178  -0.134  1.781  2.776  1.869